integration-tests/src/main/scala/za/co/absa/spline/test/ProducerModelImplicits.scala

```diff
@@ -28,20 +30,33 @@ object ProducerModelImplicits {
         .map(walker.attributeById)
     }
 
-    def precedingOp(implicit walker: LineageWalker): DataOperation =
-      walker.precedingOp(dataOperation)
+    def precedingDataOp(implicit walker: LineageWalker): DataOperation = {
+      val children = walker.precedingDataOps(dataOperation)
+      assert(children.size == 1)
+      children.head
+    }
+
+    def precedingDataOps(implicit walker: LineageWalker): Seq[DataOperation] =
+      walker.precedingDataOps(dataOperation)
+
+    def dagLeafs(implicit walker: LineageWalker): (Seq[ReadOperation], Seq[DataOperation]) =
+      findLeaves(dataOperation)
 
-    def precedingOps(implicit walker: LineageWalker): Seq[DataOperation] =
-      walker.precedingOps(dataOperation)
   }
 
   implicit class WriteOperationOps(val write: WriteOperation) extends AnyVal {
 
-    def precedingOp(implicit walker: LineageWalker): DataOperation =
-      walker.precedingOp(write)
+    def precedingDataOp(implicit walker: LineageWalker): DataOperation = {
+      val children = walker.precedingDataOps(write)
+      assert(children.size == 1)
+      children.head
```

Comment:
Here I would use a _destructuring assignment_ syntax as I think it communicates the intention better than the `assert` statement. Also because `assert` call could  be skipped at runtime nothing would prevent the rest of the values from being silently skipped should they happen to appear in the list for whatever reason.
```suggestion
      val Seq(child) = walker.precedingDataOps(write)
      child
```

Reply:
fixed